dot/peerset/peerset.go

```diff
@@ -0,0 +1,504 @@
+			peerState.discover(setId, reservedNode)
+			// we can extract discover node in ps data using peerId...
+		} else {
+			// Print Error, peer connection status is empty. May be not present in ps...
+			return
+		}
+		err := peerState.tryOutgoing(setId, reservedNode)
+		if err != nil {
+			// An error is returned only if no slot is available. Reserved nodes are
+			// marked in the state machine with a flag saying "doesn't occupy a slot",
+			// and as such this should never happen.
+			//return err
+		}
+		message := Message{
+			messageStatus: Connect,
+			setId:         uint64(setId),
+			peerId:        reservedNode,
+			//Not required for Connect and Drop...
+			IncomingIndex: 0,
+		}
+		ps.messageQueue = append(ps.messageQueue, message)
+
+	}
+	//TODO
+	// 2. Loop for fill outgoing slot if available with highest reputated not connected peer...
+
+	// Now, we try to connect to other nodes.
+	// Nothing more to do if we're in reserved mode.
+	if ps.isReservedOnly {
+		return
+	}
+
+	for {
+		if !peerState.hasFreeOutgoingSlot(setId) {
+			break
+		}
+		peerId := peerState.highestNotConnectedPeer(setId)
+		if peerId == "" {
+			break
+		}
+		node := peerState.nodes[peerId]
+		if node.getReputation() < BannedThreshold {
+			break
+		}
+		err := peerState.tryOutgoing(setId, peerId)
+		if err != nil {
+			break
+		}
+		message := Message{
+			messageStatus: Connect,
+			setId:         uint64(setId),
+			peerId:        peerId,
+			//Not required for Connect and Drop...
+			IncomingIndex: 0,
+		}
+		ps.messageQueue = append(ps.messageQueue, message)
+
+	}
+
+}
+
+func (ps *Peerset) onAddReservedPeer(setId int, peerId peer.ID) {
+	if _, ok := ps.reservedNode[peerId]; ok {
+		return
+	}
+	ps.reservedNode[peerId] = true
+	ps.data.addNoSlotNode(setId, peerId)
+	ps.allocSlot(setId)
+}
+func (ps *Peerset) onRemoveReservedPeer(setId int, peerId peer.ID) {
+	if _, ok := ps.reservedNode[peerId]; !ok {
+		return
+	}
+	state := ps.data
+	delete(ps.reservedNode, peerId)
+	state.removeNoSlotNode(setId, peerId)
+
+	// Nothing more to do if not in reserved-only mode.
+	if !ps.isReservedOnly {
+		return
+	}
+	//TODO analyse why only for reserved only mode we are disconnecting...
+
+	// reservedOnly mode is not yet implemented for future this code will help.
+	// If, however, the ps is in reserved-only mode, then the removed node needs to be
+	// disconnected.
+
+	if ps.data.peer(setId, peerId) == ConnectedPeer {
+		ps.data.disconnect(setId, peerId)
+		message := Message{
+			messageStatus: Drop,
+			setId:         uint64(setId),
+			peerId:        peerId,
+			//Not required for Connect and Drop...
+			IncomingIndex: 0,
+		}
+		ps.messageQueue = append(ps.messageQueue, message)
+	}
+
+}
+
+// TODO: Analysis and implementation yet to be done for the usage of this in gossamer...
+func (ps *Peerset) onSetReservedPeer(setId int, peerId peer.ID) {
+	return
+}
+
+// Returns the list of reserved peers.
+func (ps *Peerset) reservedPeers(setId int) []peer.ID {
```

Comment:
I think this is the same functionality as the current network "persistent peers"

Reply:
I have moved the persistent peers functionality to peerset.